### PR TITLE
client/tailscale, cmd/tailscale, localapi: add 'tailscale nc'

### PR DESCRIPTION
**Pull request by bradfitz**
This adds a "tailscale nc" command that acts a bit like "nc", but
dials out via tailscaled via localapi.

This is a step towards a "tailscale ssh", as we'll use "tailscale nc"
as a ProxyCommand for in some cases (notably in userspace mode).

But this is also just useful for debugging & scripting.

Updates #3802

RELNOTE=tailscale nc

### CODE REVIEW

File: ipn/localapi/localapi.go [hunk 1]
```diff
@@ -542,6 +546,84 @@ func (h *Handler) serveSetExpirySooner(w http.ResponseWriter, r *http.Request) {
...
+		_, err := io.Copy(outConn, reqConn)
+		errc <- err
+	}()
+	<-errc
+}
+
+// drainBufConn is a net.Conn with an initial bunch of bytes in a
+// bufio.Reader. Read drains the bufio.Reader until empty, then passes
+// through subsequent reads to the Conn directly.
+type drainBufConn struct {
```
Comment (maisem) [inline ipn/localapi/localapi.go:611]: We have this in a few places, might be time to put it in `netutil`?

File: client/tailscale/tailscale.go [hunk 2]
```diff
@@ -419,6 +420,77 @@ func SetDNS(ctx context.Context, name, value string) error {
...
+	if !ok {
+		res.Body.Close()
+		return nil, errors.New("http Transport did not provide a writable body")
+	}
+	return &wrappedConn{switchedConn, rwc}, nil
+}
+
+type wrappedConn struct {
+	net.Conn
+	rwc io.ReadWriteCloser
```
Comment (maisem) [inline client/tailscale/tailscale.go:479]: Go question, can we do:
```
type wrappedConn struct {
  net.Conn
  io.ReadWriteCloser
}

wc := wrappedConn{conn, resp.Body}
```

File: client/tailscale/tailscale.go [hunk 3]
```diff
@@ -419,6 +420,77 @@ func SetDNS(ctx context.Context, name, value string) error {
...
+	rwc, ok := res.Body.(io.ReadWriteCloser)
+	if !ok {
+		res.Body.Close()
+		return nil, errors.New("http Transport did not provide a writable body")
+	}
+	return &wrappedConn{switchedConn, rwc}, nil
+}
+
+type wrappedConn struct {
+	net.Conn
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply client/tailscale/tailscale.go:478]: there's a defer already

File: client/tailscale/tailscale.go (hunk 2 above)
Comment (bradfitz) [reply client/tailscale/tailscale.go:479]: Try: https://go.dev/play/p/Q22iddFt1WL

Comment (maisem) [reply client/tailscale/tailscale.go:479]: So it doesn't work if there is any ambiguity? 

https://go.dev/play/p/9bhuCnlGezi

File: client/tailscale/tailscale.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maisem) [reply client/tailscale/tailscale.go:478]: Where?

File: ipn/localapi/localapi.go (hunk 1 above)
Comment (bradfitz) [reply ipn/localapi/localapi.go:611]: https://github.com/tailscale/tailscale/pull/4273

**Comment by bradfitz** [comment]
PTAL

**Review by maisem — APPROVED** [review]

**Pull request merged**